#9645 Implement FTP extensions for IPv6

cjwatson opened the pull request:
Fixes #9645

These are defined in RFC 2428.

The facilities in the RFC for requesting the use of a specific network
protocol don't fit very well into Twisted's view of the world, but I've
at least arranged for the command arguments to be validated
appropriately for the common cases of IPv4 and IPv6.

Tests for IPv6 support are only partial, because FTPClient doesn't yet
support this (see https://twistedmatrix.com/trac/ticket/9085).

cjwatson committed:
Implement FTP extensions for IPv6

These are defined in RFC 2428.

The facilities in the RFC for requesting the use of a specific network
protocol don't fit very well into Twisted's view of the world, but I've
at least arranged for the command arguments to be validated
appropriately for the common cases of IPv4 and IPv6.

Tests for IPv6 support are only partial, because FTPClient doesn't yet
support this (see https://twistedmatrix.com/trac/ticket/9085).

cjwatson committed:
Skip most EPSV tests if IPv6 is unsupported

cjwatson committed:
Fix some lint

cjwatson committed:
Make EPSV tests connect to ::1, not 127.0.0.1

cjwatson commented:
There are some remaining lint errors, but they're essentially all due to the justification of "=" characters in constant assignments near the top of the file, and mostly pre-existing except where I made my changes fit the surrounding code.  I think it would make more sense to waive these rather than clutter this PR with them, but if reviewers require it then I can do the reformatting.

cjwatson committed:
Fix behaviour of PASV with IPv4-mapped addresses

When receiving a connection to an IPv6 endpoint using an IPv4-mapped
address (possible on some IPv6 stacks), PASV can work but needs to
return a properly-encoded IPv4 address.  Including the leading "::ffff:"
in the response appears to confuse connection tracking in some
firewalls.

cjwatson committed:
Fix some lint

cjwatson committed:
Test that PASV is rejected over IPv6

hawkowl committed:
Merge branch 'trunk' into 9645-ftp-rfc2428

glyph committed:
Merge branch 'trunk' into 9645-ftp-rfc2428

adiroiban commented on src/twisted/protocols/ftp.py:
```diff
@@ -737,6 +766,8 @@ class FTP(basic.LineReceiver, policies.TimeoutMixin, object):
     dtpInstance = None
     binary = True
+    epsvAll = False
+    supportedNetworkProtocols = (_AFNUM_IP, _AFNUM_IP6)
```
Do we need these members as public members?

cjwatson replied:
I suppose not.  I've made them private now.

adiroiban commented on src/twisted/protocols/ftp.py:
```diff
@@ -981,6 +1108,51 @@ def connFailed(err):
+            transport addresses.
+        """
+        if self.epsvAll:
+            return defer.fail(BadCmdSequenceError(
```
I think that is ok to return a failure here...

What I did, is create a standard reponse code like 

```
RESPONSE[EPSV_ALL_ACTIVE] = '503 EPSV ALL was enabled.'
```

and then just do

```
        if self._epsvAllActive:
            return (EPSV_ALL_ACTIVE,)
```

cjwatson replied:
I suppose I could have done, but I was following the existing pattern set by `ftp_PASV` and `ftp_PORT`, and it didn't really seem worth lifting this up to the top level when each message is only used in one place.

adiroiban commented on src/twisted/protocols/ftp.py:
```diff
@@ -945,6 +978,29 @@ def ftp_PASV(self):
+                # response in order that at least clients that ignore the
+                # host part can work, and if it becomes necessary then we
+                # could do that too.)
+                return defer.fail(CmdNotImplementedError('PASV'))
```
This is ok.

Maybe the error can be:  `PASV available only for IPv4` ... or something more verbose for the client :)

Adding support for mapped IPV4 is nice, but I don't know if it is worth adding all this extra handling here.

What I have found for many of my deployments is that with FTP servers behind the load balances, people will want to hardcode the Load Balancer IP in the PASV response... so that old clients will continue to work.

cjwatson replied:
I've made the error message clearer.

Not supporting IPv4-mapped addresses caused a [regression](https://bugs.launchpad.net/txpkgupload/+bug/1836059) in our environment, which is why I added this.  It's a relatively small amount of extra code compared to `EPSV` as a whole.

glyph commented on src/twisted/protocols/ftp.py:
```diff
@@ -706,6 +730,11 @@ class FTP(basic.LineReceiver, policies.TimeoutMixin, object):
     @ivar listenFactory: A callable with the signature of
         L{twisted.internet.interfaces.IReactorTCP.listenTCP} which will be used
         to create Ports for passive connections (mainly for testing).
+    @ivar epsvAll: If true, "EPSV ALL" was received from the client, requiring
```
Thanks so much for proactively adding docs for this!

glyph commented on src/twisted/protocols/ftp.py:
```diff
@@ -737,6 +766,8 @@ class FTP(basic.LineReceiver, policies.TimeoutMixin, object):
     dtpInstance = None
     binary = True
+    epsvAll = False
+    supportedNetworkProtocols = (_AFNUM_IP, _AFNUM_IP6)
```
Appreciate that you made this a tuple so that we wouldn't get cruft accumulating in it by accident :)

glyph commented on src/twisted/protocols/ftp.py:
```diff
@@ -864,14 +896,15 @@ def call_ftp_command(command):
 
 
-    def getDTPPort(self, factory):
+    def getDTPPort(self, factory, interface=''):
```
Can you add a `@param` explaining its usage, since this is technically modifying a public API?

Also, `''` is an ambiguous default value; is this supposed to be bytes or text?

cjwatson replied:
I've added the `@param`.  The ambiguous default was an issue when you said this, but is no longer ambiguous now that Twisted only supports Python 3.

glyph commented on src/twisted/protocols/ftp.py:
```diff
@@ -945,6 +978,29 @@ def ftp_PASV(self):
+        try:
+            address = ipaddress.IPv6Address(_coercedUnicode(host))
+        except ipaddress.AddressValueError:
+            pass
```
Minor stylistic note: the best exception-handling form here would be to be *even more* constrained here; i.e.

```
coercedHost = _coercedUnicode(host)
try:
    address = IPv6Address(coercedHost)
except ...:
    pass
```

Mainly just a note for future reference.

cjwatson replied:
Effectively done by merging current Twisted trunk, which dropped Python 2 compatibility and thus the `_coercedUnicode` call just goes away.

glyph commented on src/twisted/test/test_ftp.py:
```diff
@@ -36,8 +37,75 @@
 
 
 
+def _has_ipv6():
```
`test_conch` and `test_udp` both already define functions with *exactly* this name, and almost exactly this same implementation. 3 total repetitions of this code is too many :).  Please refactor it into a common (private, for now) location, so we're not making 3 different useless sockets during the test suite.

glyph commented on src/twisted/test/test_ftp.py:
```diff
@@ -36,8 +37,75 @@
+
+
+
+def _has_ipv4_mapped_addresses():
```
ahem, `_hasIPv4MappedAddresses` :)

glyph commented on src/twisted/test/test_ftp.py:
```diff
@@ -11,6 +11,7 @@
 import getpass
 import string
 import random
+import socket
```
For what it's worth, these tests are in the old, bad style.  The new, good style would be to consistently use `twisted.internet.testing.MemoryReactor` and not require real, actual IPv6 local connectivity in order to cover this code.  I realize that this isn't really a new problem in this branch, so it doesn't need to be addressed by this branch, but I feel like I should really bring it up to get the word out.

glyph commented on src/twisted/test/test_ftp.py:
```diff
@@ -1298,6 +1568,74 @@ def checkDownload(download):
 
 
 
+class FTPServerEprtDataConnectionTests(FTPServerPasvDataConnectionTests):
```
Can you add a test docstring explaining what this covers more precisely?

glyph commented on src/twisted/test/test_ftp.py:
```diff
@@ -1298,6 +1568,74 @@ def checkDownload(download):
+    def test_EPRTCannotConnect(self):
+        """
+        Listen on a port, and immediately stop listening as a way to find a
+        port number that is definitely closed.
```
"definitely" is a strong word here :).

glyph commented on src/twisted/test/test_ftp.py:
```diff
@@ -1298,6 +1568,74 @@ def checkDownload(download):
+        """
+        Tear down the connection.
+
+        @return: L{defer.DeferredList}
```
that fires when...?

glyph commented on src/twisted/test/test_ftp.py:
```diff
@@ -1298,6 +1568,74 @@ def checkDownload(download):
+
+    def test_EPRTCannotConnect(self):
+        """
+        Listen on a port, and immediately stop listening as a way to find a
```
Can you describe this in terms of what's being verified rather than what the setup does?  ( https://jml.io/pages/test-docstrings.html is the canonical guideline. )

glyph commented on src/twisted/python/_setup.py:
```diff
@@ -288,6 +288,7 @@ class my_build_ext(build_ext_twisted):
         "hyperlink >= 17.1.1",
         "PyHamcrest >= 1.9.0",
         "attrs >= 17.4.0",
+        "ipaddress; python_version < '3.0'",
```
Thanks for correctly updating the package metadata for this requirement.  It strikes me as slightly unfortunate that this is our first `ipaddress` import anywhere, since we consistently use `twisted.internet.address` everywhere else.  Since I'm sending this back to you for feedback, maybe *consider* adding the IPv4 mapped test to our own class?  I wouldn't say it should be a hard requirement, since we probably transitively depend on this six different ways anyway now, and it is in the stdlib, but it does stick out just a little.

If you consider it and think it's the best choice, then by all means leave it :)

cjwatson replied:
I agree this is novel, and I've considered the change you suggest.  I think if `twisted.internet.address` had any interrogation properties or methods rather than being pure data containers (with a few dunder methods), my answer might be different.  As it is, it seems really difficult to justify the effort and extra API surface of adding this to `twisted.internet.address` when there's already an adequate facility in the stdlib, especially now that we require Python 3 and so we no longer need an extra dependency for it.

So I prefer to leave this as it is, although I've dropped the setup change since `python_version < '3.0'` is now always true for Twisted.

glyph replied:
> So I prefer to leave this as it is, although I've dropped the setup change since `python_version < '3.0'` is now always true for Twisted.

always False for Twisted, I hope :)

glyph reviewed (changes requested):
The changes requested are pretty minor; for the most part this is a great change.

Regarding the lint thing: when I'm looking for reviews to do, I tend to hit the ones with green checkmarks first, so I'd probably go ahead and make the style inconsistent to appease the mechanical checks.  It's not a dealbreaker, but it'll make it easier and quicker to complete the review.  Maybe we should do a lint-ratchet reduction PR to reformat the rest of the surrounding stuff but let's not bundle those together.

adiroiban reviewed (commented):
I didn't had time to finalize the review...but  I am sending my current comments.

cjwatson commented:
I filed #1190 to clean up lint, and will return to this once that's merged.

cjwatson committed:
Merge branch 'trunk' into 9645-ftp-rfc2428

cjwatson committed:
Improve various FTP docstrings and comments

cjwatson committed:
Factor out common skip-without-IPv6 code in tests

cjwatson committed:
Fix function name

cjwatson committed:
Make new FTP attributes private

cjwatson committed:
Return clearer error message for PASV on IPv6

cjwatson committed:
Test some more EPRT parsing cases

cjwatson committed:
Move decodeExtendedAddressLine to test_ftp

It's only used by tests, so moving it saves having to worry about
coverage for it.

cjwatson commented:
@glyph @adiroiban Sorry for taking nearly two years to get round to dealing with your review comments!  I think I've dealt with pretty much everything now, apart from the `MemoryReactor` comment (which I agree with but I don't have the energy to refactor all the FTP tests at this point, and also as @glyph says it's not a new problem in this branch), and the `narrativedocs` test failures which are fixed in #1667, so this should be ready for another round of review now.

cjwatson committed:
Suppress coverage complaint on some test code

cjwatson committed:
Move pragma to correct line

glyph commented:
No need to apologize — if anything, managing to finally come back after 2 years rather than just dropping it forever is praiseworthy :).

cjwatson committed:
Turn decodeExtendedAddressLine into a test method

Another attempt to pacify coverage checks.

cjwatson committed:
Merge branch 'trunk' into 9645-ftp-rfc2428

adiroiban reviewed (approved):
Looks good. Thanks.

I think that this can be merged as it's already a big improvement over trunk.

----------

 for my project I am using a highly modified version of twisted FTP protocol with support for py2 and Implicit/Explict FTPS, x509 based authentication, and machine listing and a few other extensions . client and server side.

if such functionality is needed in twisted and someone has time to review the changes, let me know and I can look at sending them for review

adiroiban commented:
I am enabled auto-merge. It looks like  the merge is blocked as @glyph  has rejected the changes.

Once the changes are approved by @glyph the PR should be automatically merged.

Thanks

adiroiban committed:
Merge branch 'trunk' into 9645-ftp-rfc2428

adiroiban commented:
@glyph can  we merge this? Do you still think that this needs changes? Thanks!

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

glyph commented:
In the future, you can always feel free to dismiss my reviews if you've reviewed the points that I raised and they've been addressed.
